qiskit/circuit/singleton.py

```diff
@@ -0,0 +1,250 @@
+# introspection).  Instead, we use the `abc.ABC`/`abc.ABCMeta` pattern of defining a concrete middle
+# class (`SingletonGate` in the `XGate` case) that sets the metaclass, selects the overrides to be
+# applied, and has an `__init_subclass__` that applies the singleton-subclass-creation steps above.
+# The overrides are in separate classes so that _mutable_ `XGate` instances do not have them in
+# their own MROs; doing this is easier to implement, but requires all the setters and checkers to
+# dance around at runtime trying to validate whether mutating the instance is allowed.
+
+
+def _impl_new(cls, *_args, **_kwargs):
+    # __new__ for the singleton instances.
+    raise TypeError(f"cannot create '{cls.__name__}' instances")
+
+
+def _impl_init_subclass(base, overrides):
+    # __init_subclass__ for the classes that make their children singletons (e.g. `SingletonGate`)
+
+    def __init_subclass__(cls, *, create_singleton=True, **kwargs):
+        super(base, cls).__init_subclass__(**kwargs)
+        if not create_singleton:
+            return
+
+        # We need to make a new type object that pokes in the overrides into the correct
+        # place in the method-resolution order.
+        singleton_class = _SingletonMeta.__new__(
+            _SingletonMeta,
+            f"_Singleton{cls.__name__}",
+            (cls, overrides),
+            # This is a dynamically generated class so it's got no module.  The slot layout of the
+            # singleton class needs to match any layout in the base.
+            {"__module__": None, "__slots__": (), "__new__": _impl_new, "_base_class": cls},
+            create_singleton=False,
+        )
+
+        # Make a mutable instance, fully instantiate all lazy properties, then freeze it.
+        cls._singleton_instance = cls(_force_mutable=True)
+        cls._singleton_instance._define()
+        cls._singleton_instance.__class__ = singleton_class
+
+    return classmethod(__init_subclass__)
+
+
+class _SingletonMeta(type(Instruction)):
+    # The inheritance above is to ensure metaclass compatibility with `Instruction`, though pylint
+    # doesn't understand that this is a subclass of `type`, so uses metaclass `self` conventions.
+
+    # pylint: disable=bad-classmethod-argument,no-self-argument
+
+    __slots__ = ()
+
+    def __new__(mcs, name, bases, namespace, *, overrides=None, **kwargs):
+        cls = super().__new__(mcs, name, bases, namespace, **kwargs)
+        if overrides is not None:
+            # The `__init_subclass__` logic is shared between `SingletonInstruction` and
+            # `SingletonGate`, but we can't do `__init_subclass__ = _impl_init_subclass(...)`
+            # inside the class definition bodies because it couldn't make `super()` calls.
+            cls.__init_subclass__ = _impl_init_subclass(cls, overrides)
+        return cls
+
+    def __call__(cls, *args, _force_mutable=False, **kwargs):
```

Comment:
I did wonder whether we ought to promote `_force_mutable` up to a public kwarg, to allow use like `XGate(force_mutable=True)`, but I wasn't certain how useful it would be publicly, and I didn't love the name.

Reply:
I think it's fine to leave it private. Having a consistent pattern for all gates of `Gate().to_mutable()` seems totally reasonable, and also having a public `Gate(force_mutable=True)` on a subset of classes just makes it a bit confusing from an api perspective.

Reply:
Yeah, I guess if we wanted to do it properly we'd have to give `Instruction` a metaclass that just ate that kwarg.  That's too much to do without a hard compelling use case.